core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -121,10 +128,66 @@ public static <T extends ScanTask> CloseableIterable<ScanTaskGroup<T>> planTaskG
     Function<T, Long> weightFunc =
         task -> Math.max(task.sizeBytes(), task.filesCount() * openFileCost);
 
+    return toTaskGroupIterable(splitTasks, splitSize, lookback, weightFunc);
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends PartitionScanTask> List<ScanTaskGroup<T>> planTaskGroups(
+      List<T> tasks,
+      long splitSize,
+      int lookback,
+      long openFileCost,
+      Types.StructType projectedPartitionType) {
+
+    Preconditions.checkArgument(splitSize > 0, "Invalid split size (negative or 0): %s", splitSize);
+    Preconditions.checkArgument(
+        lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
+    Preconditions.checkArgument(
+        openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
+
+    List<T> splitTasks = Lists.newArrayList();
+    for (T task : tasks) {
+      if (task instanceof SplittableScanTask<?>) {
+        ((SplittableScanTask<? extends T>) task).split(splitSize).forEach(splitTasks::add);
+      } else {
+        splitTasks.add(task);
+      }
+    }
+
+    Function<T, Long> weightFunc =
+        task -> Math.max(task.sizeBytes(), task.filesCount() * openFileCost);
+
+    Map<Integer, StructProjection> projectionsBySpec = Maps.newHashMap();
+
+    // Group tasks by their partition values
+    StructLikeMap<List<T>> groupedTasks = StructLikeMap.create(projectedPartitionType);
+
+    for (T task : splitTasks) {
+      PartitionSpec spec = task.spec();
+      projectionsBySpec.computeIfAbsent(
+          spec.specId(),
+          specId -> StructProjection.create(spec.partitionType(), projectedPartitionType));
+      StructProjection projectedStruct = projectionsBySpec.get(spec.specId()).copy();
+      groupedTasks
+          .computeIfAbsent(projectedStruct.copy().wrap(task.partition()), k -> Lists.newArrayList())
+          .add(task);
+    }
+
+    // Now apply task combining within each partition
+    return groupedTasks.values().stream()
+        .flatMap(
+            ts ->
+                StreamSupport.stream(
+                    toTaskGroupIterable(ts, splitSize, lookback, weightFunc).spliterator(), false))
+        .collect(Collectors.toList());
+  }
+
+  private static <T extends ScanTask> CloseableIterable<ScanTaskGroup<T>> toTaskGroupIterable(
```

Comment:
I understand we want to reuse this method in both calls. I am afraid this will lead to a resource leak as the existing call passes a closable iterable, which wouldn't be properly closed now. Also, if we return a closable iterable, it is strange to collect it into a list without closing.

I suggested `toTaskGroupStream` to avoid weird Spotless formatting for the closure. If you feel it is easier to just have a closure, let's drop the extra method. Up to you.

Reply:
OK, let me change it to `toTaskGroupStream`.